### PR TITLE
Improved models for SIMD intrinsics

### PR DESCRIPTION
**Pull request by karthikbhargavan**
This PR does some cleanup in the SIMD intrinsics models, with a view to unifying
the proofs in SHA-3, ML-KEM, and ML-DSA and adding more tests for these models:

-`crates/utils/core-models` — executable reference models, differential tests, and lemmas
-`crates/utils/intrinsics` — the trusted F*  *axioms* in `avx2_extract.rs` / `arm64_extract.rs` 

This PR is layered on top of the upgrade-0.3.7 branch and is a pre-requisite for merging the SHA-3, ML-KEM, and ML-DSA proofs.

### CODE REVIEW

**Review by franziskuskiefer — DISMISSED** [review]
Looks like CI isn't happy.
Also, should this be against the hax-0.3.7 branch instead of main then?

**Comment by karthikbhargavan** [comment]
> Looks like CI isn't happy. Also, should this be against the hax-0.3.7 branch instead of main then?

yep, fixing.

File: crates/utils/core-models/scripts/classify-nospec.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
+#!/usr/bin/env python3
+"""
+classify-nospec.py — For each genuinely spec-free T1 wrapper, find call sites
+in ml-kem / ml-dsa / sha-3 and report the verification status of the caller.
```
Comment (franziskuskiefer) [inline crates/utils/core-models/scripts/classify-nospec.py:4]: All these things look global from the outside but only do something for those three algorithms. That needs to be made clear somewhere, and we need a way to extend this for other algorithms like AES or anything else that's using intrinsics.

File: crates/utils/core-models/src/core_arch/arm/interpretations.rs [hunk 2]
```diff
@@ -0,0 +1,1206 @@
...
+
+    pub fn vmlal_high_s16(a: i32x4, b: i16x8, c: i16x8) -> i32x4 {
+        i32x4::from_fn(|i| a[i].wrapping_add((b[4 + i] as i32) * (c[4 + i] as i32)))
+    }
+
+    /// VQDMULH (saturating doubling multiply, returning the high half).
+    /// `result[i] = saturate_to_i16( (2 * a[i] * b[i]) >> 16 )`.
+    /// Saturation only fires for the (i16::MIN, i16::MIN) corner case (which
+    /// would yield a value >= 2^15 = i16 overflow).
+    pub fn vqdmulhq_s16(a: i16x8, b: i16x8) -> i16x8 {
```
Comment (franziskuskiefer) [inline crates/utils/core-models/src/core_arch/arm/interpretations.rs:112]: This doesn't appear to handle the `(i16::MIN, i16::MIN)` case correctly.
This makes me think that there may be more issues (`vqdmulhq_n_s32` appears to have it too). Can you improve testing, esp. for corner cases and make the PR more reviewable, i.e. into smaller digestible chunks?

File: crates/utils/core-models/scripts/check-intrinsics-parity.sh [hunk 3]
```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/env bash
```
Comment (franziskuskiefer) [inline crates/utils/core-models/scripts/check-intrinsics-parity.sh:1]: This script fails. Please make sure it works and we should run it on CI to make sure it's an actual CI gate.

File: crates/utils/intrinsics/src/avx2_extract.rs [hunk 4]
```diff
@@ -587,6 +995,28 @@ pub fn mm256_sllv_epi32(vector: Vec256, counts: Vec256) -> Vec256 {
...
+  = let aux (i: nat{i < 4})
+      : Lemma (get_lane_u64x4 (mm256_slli_epi64 v_LEFT x) i ==
+               (get_lane_u64x4 x i <<! v_LEFT)) =
+      Rust_primitives.Integers.lemma_int_t_eq_via_bits
+        (get_lane_u64x4 (mm256_slli_epi64 v_LEFT x) i)
+        (get_lane_u64x4 x i <<! v_LEFT)
+    in FStar.Classical.forall_intro aux
+"#
+)]
+#[hax_lib::requires(LEFT >= 0 && LEFT <= 64)]
```
Comment (franziskuskiefer) [inline crates/utils/intrinsics/src/avx2_extract.rs]: Is it `<= 64` or `<64`. `lemma_mm256_slli_epi64_u64x4` says `<64`.

File: crates/utils/core-models/src/core_arch/x86/interpretations.rs [hunk 5]
```diff
@@ -473,11 +486,22 @@ pub mod int_vec {
...
+    // shift count (low 8 bits of IMM8) is greater than 15, the destination
+    // is zeroed. Mirrors rust-lang/stdarch#1823 (which fixed an earlier
+    // `IMM8 % 16` bug that wrapped the shift instead of zeroing).
     pub fn _mm256_bsrli_epi128<const IMM8: i32>(a: i128x2) -> i128x2 {
         i128x2::from_fn(|i| {
+            // Intel spec: a byte-shift amount > 15 clears the lane (the count is
+            // clamped to 16), it is NOT taken modulo 16. The old `tmp % 16` here
+            // mirrored a Rust core bug whose fix we upstreamed, so the model must
+            // follow the spec rather than the (now-corrected) implementation.
             let tmp = IMM8 % 256;
```
Comment (franziskuskiefer) [inline crates/utils/core-models/src/core_arch/x86/interpretations.rs]: Some are using `rem_euclid` and other like this one `% 256`. We should make sure it's consistent and what we want.

File: crates/utils/core-models/scripts/intrinsics-audit.py [hunk 6]
```diff
@@ -0,0 +1,844 @@
...
+
+    write_csv(t1, args.csv)
+    write_md(t1, t2_map, t3, args.md)
+
+    summary = emit_summary_line(t1)
+    if args.print_summary:
+        print(summary)
+    else:
+        print(summary)
+    return 0
```
Comment (franziskuskiefer) [inline crates/utils/core-models/scripts/intrinsics-audit.py]: This seems a bit silly.

File: crates/utils/core-models/proofs/intrinsics-trust-index.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
+# SIMD intrinsics trust index
+
+Generated by `crates/utils/core-models/scripts/intrinsics-audit.py`.
+See `crates/utils/core-models/INTRINSICS-TRUST-PLAN.md` for the trust-ladder definition (L0..L4) and the D6 sub-percentages.
```
Comment (franziskuskiefer) [inline crates/utils/core-models/proofs/intrinsics-trust-index.md:4]: There's no `INTRINSICS-TRUST-PLAN.md`.

File: fstar-helpers/fstar-bitvec/Bitvec.U64Rotate.fst [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ *)
+module Bitvec.U64Rotate
+
+open Core_models
+
+let in_range64 (sh: u32) : prop =
+  Rust_primitives.Integers.v sh > 0 /\ Rust_primitives.Integers.v sh < 64
+
+(* Bridge lemma — see module header. *)
+assume
```
Comment (franziskuskiefer) [inline fstar-helpers/fstar-bitvec/Bitvec.U64Rotate.fst:26]: I don't think this is listed as axiom?

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
I think this needs a bit of work/cleanup/splitup.

The checked in F* code is stale. Should it be dropped?

**Comment by jschneider-bensch** [comment]
I'd like to change the base to branch [`jonas/upgrade-hax-lib-0.3.7`](https://github.com/celabshq/libcrux/tree/jonas/upgrade-hax-lib-0.3.7), but I'm not yet doing it because there are open review conversations that may get lost as a result.
@karthikbhargavan Are you planning on working on these still?

**Comment by karthikbhargavan** [comment]
Yes, I'll do a new round of changes that you can review on Monday, and then rebase.,

File: crates/utils/core-models/src/core_arch/x86/interpretations.rs (hunk 5 above)
Comment (karthikbhargavan) [reply crates/utils/core-models/src/core_arch/x86/interpretations.rs]: you are right, will fix.

File: crates/utils/intrinsics/src/avx2_extract.rs (hunk 4 above)
Comment (karthikbhargavan) [reply crates/utils/intrinsics/src/avx2_extract.rs]: you are right, will fix.

File: crates/utils/core-models/src/core_arch/arm/interpretations.rs (hunk 2 above)
Comment (karthikbhargavan) [reply crates/utils/core-models/src/core_arch/arm/interpretations.rs:112]: you are right will fix.

**Comment by karthikbhargavan** [comment]
> I think this needs a bit of work/cleanup/splitup.
> 
> The checked in F* code is stale. Should it be dropped?

I am not sure which code you think is stale?

File: crates/utils/core-models/scripts/classify-nospec.py (hunk 1 above)
Comment (karthikbhargavan) [reply crates/utils/core-models/scripts/classify-nospec.py:4]: This is a best-effort TCB tracking script which may indeed be too specific. Maybe it should be dropped in favor of some more general libcrux tracking scripts?

File: crates/utils/core-models/scripts/check-intrinsics-parity.sh (hunk 3 above)
Comment (karthikbhargavan) [reply crates/utils/core-models/scripts/check-intrinsics-parity.sh:1]: Hm, something broke maybe, will check,

File: crates/utils/core-models/scripts/intrinsics-audit.py (hunk 6 above)
Comment (karthikbhargavan) [reply crates/utils/core-models/scripts/intrinsics-audit.py]: ha.

File: fstar-helpers/fstar-bitvec/Bitvec.U64Rotate.fst (hunk 8 above)
Comment (karthikbhargavan) [reply fstar-helpers/fstar-bitvec/Bitvec.U64Rotate.fst:26]: Hm, will check,

**Comment by karthikbhargavan** [comment]
I think most of the comments are now addressed and we added more tests and proofs. But, the one thing that would be useful to do is to divide up the work in reviewable chunks. Would it help if I reorganized the commits so that each commit brings together one set of intrinsics, and you folks can review them commit-by-commit? This might make the comments above obsolete though, so if we decide to do that, we should make sure the comments above are satisfactorily resolved.

**Comment by karthikbhargavan** [comment]
@jschneider-bensch do you want to rebase this on your 0.3.7 branch/PR and then I'll redo the commits to make them reviewable in smaller chunks?

**Comment by abentkamp** [comment]
I've rebased on `main` (which includes Jonas's 0.3.7 PR).

File: crates/utils/core-models/scripts/classify-nospec.py (hunk 1 above)
Comment (abentkamp) [reply crates/utils/core-models/scripts/classify-nospec.py:4]: Karthik extended the script to also cover AES and kmac. Maybe the script output should say exactly which crates have been searched?

File: crates/utils/core-models/src/core_arch/arm/interpretations.rs (hunk 2 above)
Comment (abentkamp) [reply crates/utils/core-models/src/core_arch/arm/interpretations.rs:112]: Karthik fixed the bug and added corner-case tests for various functions, both against the chip and host-independent.

File: crates/utils/intrinsics/src/avx2_extract.rs (hunk 4 above)
Comment (abentkamp) [reply crates/utils/intrinsics/src/avx2_extract.rs]: Fixed.

File: crates/utils/core-models/src/core_arch/x86/interpretations.rs (hunk 5 above)
Comment (abentkamp) [reply crates/utils/core-models/src/core_arch/x86/interpretations.rs]: Fixed.

File: crates/utils/core-models/scripts/intrinsics-audit.py (hunk 6 above)
Comment (abentkamp) [reply crates/utils/core-models/scripts/intrinsics-audit.py]: Fixed.

File: crates/utils/core-models/proofs/intrinsics-trust-index.md (hunk 7 above)
Comment (abentkamp) [reply crates/utils/core-models/proofs/intrinsics-trust-index.md:4]: Karthik committed the `INTRINSICS-TRUST-PLAN.md`. I am not sure it is very helpful in this form though. I'd prefer a readme by humans for humans :sweat_smile:.

File: fstar-helpers/fstar-bitvec/Bitvec.U64Rotate.fst (hunk 8 above)
Comment (abentkamp) [reply fstar-helpers/fstar-bitvec/Bitvec.U64Rotate.fst:26]: This is now mentioned in `crates/utils/core-models/proofs/intrinsics-trust-index.md`.

**Comment by abentkamp** [comment]
> I think this needs a bit of work/cleanup/splitup.
> 
> The checked in F* code is stale. Should it be dropped?

@franziskuskiefer Could you clarify which F* code you mean?

**Comment by franziskuskiefer** [comment]
> @franziskuskiefer Could you clarify which F* code you mean?

The F* code that's in the core-models directory is not an extraction of the Rust code that's in there.

**Comment by karthikbhargavan** [comment]
I am not very happy with the trust story here and we can do better. Maybe this needs another round.
